File: app/src/main/java/org/evcs/android/model/Location.java
package org.evcs.android.model;

import com.google.android.gms.maps.model.LatLng;
import com.google.maps.android.clustering.ClusterItem;

import org.jetbrains.annotations.Nullable;

import java.io.Serializable;
import java.util.List;
import java.util.Objects;

public class Location implements Serializable, ClusterItemWithText {
    public int id;
    public String name;
    public double latitude;
    public double longitude;
    public List<String> imageUrls;
/*        "station_count": {
        "dc50kw": 4,
                "dc": 0,
                "ac": 0
    },*/
    public Float distance;
    public List<Station> stations;
    public Address address;
//       "directions":"TO BE DEFINED",
//               "open_247":true,
//               "authority":"public",

//        "noodoe_id": "abc123",
//        "contact": {
//        "name": "Rina",
//        "title": "XS",
//        "email": "rina@sawayama.com",
//        "phone": "1 222 333 4444",
//        "company": "Sawayama"
//        },
//        "subscriptions_enabled": false

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Location location = (Location) o;
        return id == location.id;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    public LatLng getLatLng() {
        return new LatLng(latitude, longitude);
    }

    @Override
    public LatLng getPosition() {
        return getLatLng();
    }

    @Override
    public String getMarkerText() {
        return Integer.toString(stations.size());
    }

    public final CharSequence getPrintableDistance() {
        return (distance == null) ? "-- mi" : String.format("%.1f mi", distance);
    }
}

File: app/src/main/java/org/evcs/android/model/Location.kt
package org.evcs.android.model

import com.google.android.gms.maps.model.LatLng
import org.evcs.android.features.map.clustermap.ClusterItemWithDisabling
import org.evcs.android.features.map.clustermap.ClusterItemWithText
import org.evcs.android.features.map.clustermap.ClusterItemWithValue
import java.io.Serializable
import java.util.*

class Location : Serializable, ClusterItemWithText, ClusterItemWithValue, ClusterItemWithDisabling {
    var id = 0
    var name: String? = null
    var latitude = 0.0
    var longitude = 0.0
    var imageUrls: List<String>? = null
    var stationCount: StationCount? = null
    var distance: Float? = null
    var stations: List<Station>? = null
    var outlets: List<Outlet>? = null
    var address: Address? = null
    var comingSoon: Boolean? = null
    var gatecode: String? = null
    var directions: String? = null
//        get() = 1000


    //               "open_247":true,
    //               "authority":"public",
    //        "noodoe_id": "abc123",
    //        "contact": {
    //        "name": "Rina",
    //        "title": "XS",
    //        "email": "rina@sawayama.com",
    //        "phone": "1 222 333 4444",
    //        "company": "Sawayama"
    //        },
    //        "subscriptions_enabled": false
    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val location = o as Location
        return id == location.id
    }

    override fun hashCode(): Int {
        return Objects.hash(id)
    }

    val latLng: LatLng
        get() = LatLng(latitude, longitude)

    override fun getPosition(): LatLng {
        return latLng
    }

    override fun getMarkerText(): String {
        return stationCount!!.total().toString()
    }

    override fun getMarkerValue(): Int {
        return stationCount!!.total()
    }

    override fun isMarkerEnabled(): Boolean {
        return !(comingSoon ?: false)
    }

    override fun isClusterEnabled(other: Boolean): Boolean {
        return isMarkerEnabled() || other
    }

    val printableDistance: CharSequence
        get() = if (distance == null) "-- mi" else String.format("%.1f mi", distance)

//    val connectorTypes: Set<ConnectorType>
//        get() = stations!!.flatMap { station -> station.connectorTypes }.toSet()

    val connectorTypes: Set<ConnectorType>
        get() = outlets!!.connectorTypes()

    val acPrice: Float
        get() = priceOfFirstStationMatching { station ->
            station.getChargerType() == Station.ChargerType.AC }

    val dcPrice: Float
        get() = priceOfFirstStationMatching { station ->
            station.getChargerType() == Station.ChargerType.DC }

    fun priceOfFirstStationMatching(condition: (Station) -> Boolean) : Float {
        return stations!!.filter(condition).getOrNull(0)?.pricing?.detail?.priceKwh ?: 0f
    }

}